contrib/dhcp_agent/dhcp_agent.py

```diff
@@ -0,0 +1,438 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+import atexit
+import contextlib
+import errno
+import fcntl
+import hashlib
+import logging
+import os
+import subprocess
+import sys
+import tempfile
+from optparse import OptionParser
+from logging import handlers as logging_handlers
+
+IS_PY3 = sys.version_info[0] == 3
+if IS_PY3:
+    from dbm import gnu as gdbm
+    from urllib.request import urlopen, Request
+    from urllib.parse import urlencode
+    from urllib.error import HTTPError
+else:
+    import gdbm
+    from urllib import urlencode
+    from urllib2 import urlopen, Request, HTTPError
+
+
+APP_DIR = os.path.expanduser('~/.ralph-dhcp-agent')
+
+PROTOS = ('http', 'https')
+PROTO_HTTP, PROTO_HTTPS = PROTOS
+
+CACHE_LAST_MODIFIED_PREFIX = 'http-last-modified'
+DEFAULT_DHCP_SERVICE_NAME = 'isc-dhcp-server'
+
+
+@contextlib.contextmanager
+def open_file_or_stdout_to_writing(filename=None):
+    """Context manager open file or stdout and returns handle to file."""
+    if filename and filename in ['-', None]:
+        handler = open(filename, 'w')
+    else:
+        handler = sys.stdout
+    try:
+        yield handler
+    finally:
+        if handler is not sys.stdout:
+            handler.close()
+
+
+class Cache(object):
+    def __init__(self, cache_path):
+        if not os.path.exists(cache_path):
+            os.makedirs(cache_path)
+        self._cache = gdbm.open(os.path.join(cache_path, 'cache'), 'c')
+
+    def get(self, key, prefix):
+        url_hash = get_url_hash(key)
+        try:
+            last = self._cache[prefix + url_hash]
+        except KeyError:
+            last = None
+        return last
+
+    def set(self, key, prefix, value):
+        print(value)
+        url_hash = get_url_hash(value)
+        self._cache[prefix + url_hash] = value
+
+    def close(self):
+        self._cache.close()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args, **kwargs):
+        self.close()
+
+
+def get_url_hash(url):
+    hash_url = hashlib.md5()
+    hash_url.update(url.encode())
+    return hash_url.hexdigest()
+
+
+def get_request_params(params):
+    request_params = []
+    for key, value_list in params.items():
+        if value_list:
+            request_params.extend([(key, param) for param in value_list])
```

Comment:
Do we have a guarantee that `value_list` will always be a list and not e.g. a string..?